File: 2020/day7.py
# Cheat
# keywords: recursion, strip from right of string, string not in

nesting_rules = {}

with open('day7.tst') as f:
  for line in f:
    b, a = line.strip('\n.').split('contain')
    print("b:", b, "a:", a)
    outer = b.rstrip('bags ')
    inners = [b.strip("bags ").split(' ', 1) for b in a.split(',') if not 'no' in b]

    nesting_rules[outer] = inners

print("nesting_rules:", nesting_rules)

# part 1
def can_contain_shiny_gold(bag_colour):
    if 'shiny gold' in bag_colour:
        return True
    else:
        return any(can_contain_shiny_gold(color) for _, color in nesting_rules[bag_colour])

def number_of_bags_contained(bag_color):
    if not nesting_rules[bag_color]:
        return 0
    else:
        return sum(int(num) + int(num) * number_of_bags_contained(color) for num, color in nesting_rules[bag_color])


print(sum(can_contain_shiny_gold(color) for color in nesting_rules.keys() if color != 'shiny gold'))
print(number_of_bags_contained('shiny gold'))

File: 2020/day7.rb
# keywords: hash, match, gsub, recursion

bags = {}
contents = {}

File.open('day7.tst').readlines().each do | line|
  line = line.chomp.gsub(/ bags?|[.]/,'')
  l, r = line.split(' contain ')
  bags[l] = []

  # puts "l:#{l}, r:#{r}"
  
  r.split(",").each do |b|
    p = b.match(/ ?(\d+) (\w+ \w+)/)
    puts "#{b} | #{p[1]} , #{p[2]}" unless p.nil?
    bags[l] << [p[1].to_i, p[2]] unless p.nil?
  end
end

puts bags

def bagValue(bag, bags)
  puts "b: #{b}"

  if b.include?('shiny gold')
    return q
  else
    bags[b].each do |o|
      nn = bagValue(o, , bags)
      puts "#{spaces}> nn: #{nn}"
    end
  end
end

bags.each do |k, v|
  puts "> k:#{k}, v:#{v}"

  contents[k] = []
  v.each do |bag, q|
    n = bagValue(bag, q, bags)
    puts "> n: #{n}"
    contents[k] << [o,n]
  end

  exit()
end

contents.each { |c| pp c }